plugins/modules/s3_object_info.py

```diff
@@ -0,0 +1,570 @@
+        elif key == 'object_retention':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_retention(connection, module, bucket_name, object_name)
+        elif key == 'object_tagging':
+            all_facts[key] = {}
+            all_facts[key] = describe_s3_object_tagging(connection, module, bucket_name, object_name)
+
+    return all_facts
+
+
+def get_object(connection, module, bucket_name, object_name):
+    params = {}
+    params['Bucket'] = bucket_name
+    params['Key'] = object_name
+
+    object_info = {}
+
+    try:
+        object_info = connection.get_object(**params)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        pass
+
+    if len(object_info) != 0:
+        # Remove Body, ResponseMetadata from object_info, convert to snake_case
+        del(object_info['Body'])
+        del(object_info['ResponseMetadata'])
+        object_info = camel_dict_to_snake_dict(object_info)
+
+    return object_info
+
+
+def main():
+
+    argument_spec = dict(
+        object_details=dict(type='dict', options=dict(
+            object_acl=dict(type='bool', default=False),
+            object_attributes=dict(type='bool', default=False),
+            object_legal_hold=dict(type='bool', default=False),
+            object_lock_configuration=dict(type='bool', default=False),
+            object_retention=dict(type='bool', default=False),
+            object_tagging=dict(type='bool', default=False),
+        )),
+        bucket_name=dict(required=True, type='str'),
+        object_name=dict(type='str'),
+        object_attributes=dict(type='list', elements='str', choices=['ETag', 'Checksum', 'ObjectParts', 'StorageClass', 'ObjectSize']),
+    )
+
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+    )
+
+    try:
+        connection = module.client('s3', retry_decorator=AWSRetry.jittered_backoff())
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, msg='Failed to connect to AWS')
+
+    result = {}
+
+    bucket_name = module.params.get('bucket_name')
+    object_name = module.params.get('object_name')
+    requested_details = module.params.get('object_details')
+
+    if requested_details and requested_details['object_attributes'] is True:
+        if not module.params.get('object_attributes'):
+            module.fail_json(msg='Please provide object_attributes list to retrieve s3 object_attributes.')
+
+    if requested_details:
+        result['object_info'] = get_object_details(connection, module, bucket_name, object_name, requested_details)
+    else:
+        # if specific details are not requested, return object metadata
+        result['object_info'] = get_object(connection, module, bucket_name, object_name)
+
+    module.exit_json(msg="Retrieved s3 object info ", **result)
```

Comment:
Probably the msg seems useless to me, but I don't have a strong opinion about it.